File: path_finding/width_search.py
import numpy as np
import matplotlib.pyplot as plt
import imageio
import shutil
import os

def width_search(grid, start, end):
	"""
	Does a width-first path-finding search. Assumes only up/down/left/right moves, of equal weight.
	{{grid}}    = (n,m) integer array. -1 for impassable, 0 for passable, 1 for passed, 2 for current,
				   3 for start, 4 for end.
	{{start}}   = Start coordinates, array shape (2,)
	{{end}}     = Target coordinates, array shape (2,)
	{{return}}  = nr of moves, plus True/False if successful or not.
	"""
	np.save("grid.npy", grid)
	grid_size = np.shape(grid)
	moves_array = np.zeros( shape=(grid_size[0]*grid_size[1], 2), dtype=int )
	moves_list_broad = []
	possible_steps = np.array([[-1,0],[0,1],[1,0],[0,-1]])
	def execute_width_search():
		end_found = False
		current_tiles = np.array([start.copy()])
		new_tiles = [0]  # Just making sure new_tiles isn't empty(see next line).
		nr_moves = 0
		while np.size(new_tiles) != 0:  # While not stuck.
			new_tiles = []
			for current_tile in current_tiles:  # For every former tile,
				for step in possible_steps:     # take a step in every direction.
					tile = current_tile + step
					if 0 <= tile[0] < grid_size[0] and 0 <= tile[1] < grid_size[1]\
					and grid[tile[0],tile[1]] in [0,4]:  # Checking if passable and not out of bounds.
						if (tile == end).all():
							end_found = True
							print("You have arrived at your destination!")
							return(nr_moves, True)
						new_tiles.append(tile)
						nr_moves += 1
						grid[tile[0],tile[1]] = 1  # Marking as passed.
						moves_array[nr_moves-1] = tile
			current_tiles = np.array(new_tiles)
			moves_list_broad.append(current_tiles)
		return(nr_moves, False)  # If stuck.
	nr_moves, success = execute_width_search()
	moves_array = moves_array[:nr_moves]
	shortest_path = []
	current_tile = end
	for i in range(len(moves_list_broad)-1, -1, -1):
		for tile in moves_list_broad[i]:
			for possible_step in current_tile + possible_steps:
				if (tile == possible_step).all():
					shortest_path.append(tile)
					current_tile = tile
					break
	shortest_path = np.array(shortest_path)
	np.save("moves.npy", moves_array)
	np.save("shortest_path.npy", shortest_path)
	np.save("path_broad.npy", moves_list_broad)
	return(grid, moves_array, nr_moves, success)

File: path_finding/width_search.jl
grid = zeros(Int16, 10,10)
for i in 1:10, j in 1:10
	if rand() > 0.7
		grid[i,j] = 1
	end
end

beginning = [2, 2]
target = Int16[9, 9]
println(target)
grid[2,2] = 4
grid[9,9] = 5
println(grid)

legal_moves = Int16[-1 0 1  0;
					 0 1 0 -1]

new_tiles = zeros(Int16, 2, 30)  # "Long enough" array, to be filled with the next tiles.
current_tiles = zeros(Int16, 2, 30)
current_tiles[:,1] = beginning
new_tiles[:,1] = [-1,-1]
nr_moves = 0
nr_current_tiles = 1
while current_tiles[:,1] != [0,0]
	nr_new_tiles = 0
	for i in 1:size(legal_moves,2)
		for j in 1:nr_current_tiles
			tile = legal_moves[:,i] + current_tiles[:,j]
			if 0 < tile[1] <= 10 && 0 < tile[2] <= 10 && grid[tile[1],tile[2]] in [0 5]
				nr_moves += 1
				nr_new_tiles += 1
				new_tiles[:, nr_new_tiles] = tile
				grid[tile[1],tile[2]] = 2
				println("new tile ", tile)
				if tile == target
					println("Search Successful after $nr_moves moves!")
				end
			end
		end
	end
	current_tiles = copy(new_tiles)
	new_tiles[:] = 0
	nr_current_tiles = nr_new_tiles
end